src/DataDAOExample.sol

```diff
@@ -0,0 +1,111 @@
+
+
+contract MockMarket {
+
+    DataDAO client;
+
+    constructor(address _client) {
+        client = DataDAO(_client);
+    }
+
+    function publish_deal(bytes calldata raw_auth_params, uint256 proposalID) public {
+        // calls standard filecoin receiver on message authentication api method number
+        client.handle_filecoin_method(0, 2643134072, raw_auth_params, proposalID);
+    }
+
+}
+
+contract DataDAO {
+
+    uint64 constant public AUTHORIZE_MESSAGE_METHOD_NUM = 2643134072; 
+
+    mapping(bytes => bool) public cidSet;
+    mapping(bytes => uint) public cidSizes;
+    mapping(bytes => mapping(bytes => bool)) public cidProviders;
+
+    uint256 public proposalCount;
+
+    struct Proposal {
+        uint256 proposalID;
+        address storageProvider;
+        bytes cidraw;
+        uint size;
+        uint256 upVoteCount;
+        uint256 downVoteCount;
+        uint256 proposedAt;
+        uint256 proposalExpireAt;
+    }
+
+    mapping(uint256 => Proposal) public proposals;
+    mapping(address => mapping(uint256 => bool)) public hasVotedForProposal;
+
+    function getSP(uint256 proposalID) view internal returns(address) {
+        return proposals[proposalID].storageProvider;
+    }
+
+    function isCallerSP(uint256 proposalID) view internal returns(bool) {
+       return getSP(proposalID) == msg.sender;
+    }
+
+    function isVotingOn(uint256 proposalID) view internal returns(bool) {
+       return proposals[proposalID].proposalExpireAt > block.timestamp;
+    }
+
+    function createAddCIDProposal(bytes calldata cidraw, uint size) public {
+        proposalCount = proposalCount + 1;
+        Proposal memory proposal = Proposal(proposalCount, msg.sender, cidraw, size, 0, 0, block.timestamp, block.timestamp + 5 minutes);
+        proposals[proposalCount] = proposal;
+        cidSet[cidraw] = true;
+        cidSizes[cidraw] = size;
+    }
+
+    function upvoteCIDProposal(uint256 proposalID) public {
+        require(!isCallerSP(proposalID), "Storage Provider cannot vote his own proposal");
+        require(!hasVotedForProposal[msg.sender][proposalID], "Already Voted");
+        require(isVotingOn(proposalID), "Voting Period Finished");
+        proposals[proposalID].upVoteCount = proposals[proposalID].upVoteCount + 1;
+        hasVotedForProposal[msg.sender][proposalID] = true;
+    }
+
+    function downvoteCIDProposal(uint256 proposalID) public {
+        require(!isCallerSP(proposalID), "Storage Provider cannot vote his own proposal");
+        require(!hasVotedForProposal[msg.sender][proposalID], "Already Voted");
+        require(isVotingOn(proposalID), "Voting Period Finished");
+        proposals[proposalID].downVoteCount = proposals[proposalID].downVoteCount + 1;
+        hasVotedForProposal[msg.sender][proposalID] = true;
+    }
+     
+    function policyOK(uint256 proposalID) internal view returns (bool) {
+        //require(proposals[proposalID].proposalExpireAt > block.timestamp, "Voting in On");
```

Comment:
I think this should be `<` so that we only allow proposals that have expired their voting phase to pass the policy check. I think this prevents a single upvoter from rushing through a proposal.

But for now delete this pseudo code and if you want to indicate this as a future direction spell it out more clearly in a comment.